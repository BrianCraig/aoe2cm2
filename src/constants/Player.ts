enum Player {
    HOST = 'HOST',
    GUEST = 'GUEST',
    NONE = 'NONE'
}

export default Player;