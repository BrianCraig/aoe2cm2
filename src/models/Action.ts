enum Action {
    PICK = 'PICK',
    EXCLUSIVE_PICK = 'EXCLUSIVE_PICK',
    GLOBAL_PICK = 'GLOBAL_PICK',
    HIDDEN_PICK = 'HIDDEN_PICK',
    HIDDEN_EXCLUSIVE_PICK = 'HIDDEN_EXCLUSIVE_PICK',
    BAN = 'BAN',
    EXCLUSIVE_BAN = 'EXCLUSIVE_BAN',
    GLOBAL_BAN = 'GLOBAL_BAN',
    HIDDEN_BAN = 'HIDDEN_BAN',
    HIDDEN_EXCLUSIVE_BAN = 'HIDDEN_EXCLUSIVE_BAN',
    HIDDEN_GLOBAL_BAN = 'HIDDEN_GLOBAL_BAN',
    SNIPE = 'SNIPE',
    HIDDEN_SNIPE = 'HIDDEN_SNIPE',
    
    // REVEAL_PICKS = 'REVEAL_PICKS',
    // REVEAL_BANS = 'REVEAL_BANS',
    // REVEAL_SNIPES = 'REVEAL_SNIPES',
    REVEAL_ALL = 'REVEAL_ALL'
}

export default Action;