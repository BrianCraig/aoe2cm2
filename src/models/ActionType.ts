import Action from "./Action";

enum ActionType {
    PICK = "pick",
    BAN = "ban",
    SNIPE = "snipe",
    REVEAL = "reveal",
    NOTHING = "nothing"
}

export function actionTypeFromAction(action: Action) {
    switch (action) {
        case Action.PICK:
        case Action.EXCLUSIVE_PICK:
        case Action.GLOBAL_PICK:
        case Action.HIDDEN_PICK:
        case Action.HIDDEN_EXCLUSIVE_PICK:
            return ActionType.PICK;
        case Action.BAN:
        case Action.EXCLUSIVE_BAN:
        case Action.GLOBAL_BAN:
        case Action.HIDDEN_BAN:
        case Action.HIDDEN_EXCLUSIVE_BAN:
        case Action.HIDDEN_GLOBAL_BAN:
            return ActionType.BAN;
        case Action.SNIPE:
        case Action.HIDDEN_SNIPE:
            return ActionType.SNIPE;
        // case Action.REVEAL_PICKS:
        // case Action.REVEAL_BANS:
        // case Action.REVEAL_SNIPES:
        case Action.REVEAL_ALL:
            return ActionType.REVEAL;
    }
}

export default ActionType;