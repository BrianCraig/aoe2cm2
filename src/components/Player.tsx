import * as React from 'react';
import ActionType from "../models/ActionType";
import {default as ModelPlayer} from '../models/Player';
import {default as ModelPreset} from '../models/Preset';
import '../pure-min.css'
import '../style2.css'
import CivPanel from "../containers/CivPanel";
import Civilisation from "../models/Civilisation";
import {DraftEvent} from "../models/DraftEvent";
import PlayerEvent from "../models/PlayerEvent";

interface IProps {
    preset: ModelPreset;
    player: ModelPlayer;
    name: string;
    events?: DraftEvent[];
}

interface IState {
    bannedCivs: Civilisation[];
    pickedCivs: Civilisation[];
    snipedCivs: Civilisation[];
}

class Player extends React.Component<IProps, IState> {

    constructor(props: IProps) {
        super(props);
        this.state = eventsToState(props.events, props.player);
    }

    public componentWillReceiveProps(nextProps: Readonly<IProps>, nextContext: any): void {
        console.log("Propping up", nextProps);
        this.setState(eventsToState(nextProps.events, nextProps.player));
    }

    public render() {
        const playerId: string = 'player-' + this.props.player.toString().toLowerCase();

        const pickPanels = this.props.preset.turns.filter(turn => {
            return turn.player === this.props.player && turn.action.toString().includes('PICK');
        }).map((turn, index) => {
            let pickedCiv;
            if (this.state.pickedCivs.length > index) {
                pickedCiv = this.state.pickedCivs[index];
            }
            return React.createElement(CivPanel, {actionType: ActionType.PICK, active: false, civilisation: pickedCiv});
        });
        const banPanels = this.props.preset.turns.filter(turn => {
            return turn.player === this.props.player && turn.action.toString().includes('BAN');
        }).map((turn, index) => {
            let bannedCiv;
            if (this.state.bannedCivs.length > index) {
                bannedCiv = this.state.bannedCivs[index];
            }
            return React.createElement(CivPanel, {actionType: ActionType.BAN, active: false, civilisation: bannedCiv});
        });

        return (
            <div className="pure-u-1-2">
                <div id={playerId} className="double-outer-border">
                    <div className="double-inner-border">
                        <div className="player">
                            <div className="head-text">
                                {this.props.player}
                            </div>
                            <div className="player-head">
                                <div className="player-name">{this.props.name}</div>
                            </div>
                            <div className="chosen">
                                <div className="head-text">Picks</div>
                                <div className="picks">
                                    {pickPanels}
                                </div>

                                <div className="head-text">Bans</div>
                                <div className="bans">
                                    {banPanels}
                                </div>
                            </div>

                        </div>
                    </div>
                </div>
            </div>
        );
    }
}


function eventsToState(events: DraftEvent[] | undefined, player: ModelPlayer): IState {
    if (events === undefined) {
        return {bannedCivs: [], pickedCivs: [], snipedCivs: []};
    } else {
        events = events as DraftEvent[];
    }
    const playerEvents: PlayerEvent[] = events
        .filter(e => e instanceof PlayerEvent)
        .map(e => e as PlayerEvent)
        .filter(e => e.player === player);
    const picks: Civilisation[] = [];
    const bans: Civilisation[] = [];
    const snipes: Civilisation[] = [];
    for (const event of playerEvents) {
        switch (event.actionType) {
            case ActionType.PICK:
                picks.push(event.civilisation);
                break;
            case ActionType.BAN:
                bans.push(event.civilisation);
                break;
            case ActionType.SNIPE:
                snipes.push(event.civilisation);
                break;
        }
    }
    return {bannedCivs: bans, pickedCivs: picks, snipedCivs: snipes};
}

export default Player;
