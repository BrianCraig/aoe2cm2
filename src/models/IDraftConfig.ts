import Player from "./Player";
import {DraftEvent} from "./DraftEvent";

export interface IDraftConfig {
    nameHost: string,
    nameGuest: string,
    yourPlayerType: Player,
    events: DraftEvent[],
    nextAction: number
}