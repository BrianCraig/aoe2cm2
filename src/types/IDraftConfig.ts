import Player from "../constants/Player";
import {IDraftState} from "./index";

export interface IDraftConfig extends IDraftState {
    yourPlayerType: Player
}