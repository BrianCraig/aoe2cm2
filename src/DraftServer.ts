import express from "express";
import {Server} from "http"
import socketio from "socket.io";
import Player from "./models/Player";
import {IDraftConfig} from "./models/IDraftConfig";
import {IJoinMessage} from "./models/IJoinMessage";
import {DraftsStore} from "./models/DraftsStore";
import {Validator} from "./models/Validator";
import {ValidationId} from "./models/ValidationId";
import {DraftEvent} from "./models/DraftEvent";
import {Util} from "./models/Util";
import {AddressInfo} from "net";
import Preset from "./models/Preset";
import {Listeners} from "./models/Listeners";

export const DraftServer = {
    serve(port: string | number | undefined):{ httpServerAddr: AddressInfo | string | null; io: SocketIO.Server; httpServer: Server } {
        const app = express();
        app.set("port", port);
        app.use(express.json());

        const server = new Server(app);
        const io = socketio(server, {cookie: false});
        const draftsStore = new DraftsStore();
        const validator = new Validator(draftsStore);

        function setPlayerName(draftId: string, player: Player, name: string) {
            if (!draftsStore.has(draftId)) {
                draftsStore.initDraft(draftId, Preset.SAMPLE);
            }
            draftsStore.setPlayerName(draftId, player, name);
        }

        app.post('/preset/new', (req, res) => {
            console.log('/preset/new', req.body);
            let draftId = Util.newDraftId();
            while(draftsStore.has(draftId)){
                draftId += Util.randomChar();
            }
            const pojo: Preset = req.body.preset as Preset;
            let preset = Preset.fromPojo(pojo);
            if (preset !== undefined) {
                draftsStore.initDraft(draftId, preset);
            }
            res.json({draftId});
        });
        app.use('/draft/[a-zA-Z]+', (req, res) => {
            res.sendFile(__dirname + '/index.html');
        });

        app.use(express.static('build'));
        app.use('/', (req, res) => {
            res.sendFile(__dirname + '/index.html');
        });

        io.on("connection", (socket: socketio.Socket) => {
            const draftId: string = socket.handshake.query.draftId;

            const roomHost: string = `${draftId}-host`;
            const roomGuest: string = `${draftId}-guest`;
            const roomSpec: string = `${draftId}-spec`;

            console.log("a user connected to the draft", draftId);

            if (!draftsStore.has(draftId)) {
                socket.emit('message', 'This draft does not exist.');
                socket.disconnect(true);
                return;
            }

            const {nameHost, nameGuest} = draftsStore.getPlayerNames(draftId);

            const rooms = Object.keys(socket.adapter.rooms);
            console.log('rooms', rooms);
            if (!rooms.includes(roomHost)) {
                socket.join(roomHost);
                const message = {nameHost, nameGuest, youAre: Player.HOST};
                console.log('sending', message);
            } else if (!rooms.includes(roomGuest)) {
                socket.join(roomGuest);
                const message = {nameHost, nameGuest, youAre: Player.GUEST};
                console.log('sending', message);
            } else {
                socket.join(roomSpec);
                const message = {nameHost, nameGuest, youAre: Player.NONE};
                console.log('sending', message);
            }

            socket.on("join", (message: IJoinMessage, fn: (dc: IDraftConfig) => void) => {
                console.log("player joined:", message);
                let playerType: Player = Player.NONE;
                if (Object.keys(socket.rooms).includes(roomHost)) {
                    setPlayerName(draftId, Player.HOST, message.name);
                    draftsStore.setPlayerReady(draftId, Player.HOST);
                    playerType = Player.HOST;
                } else if (Object.keys(socket.rooms).includes(roomGuest)) {
                    setPlayerName(draftId, Player.GUEST, message.name);
                    draftsStore.setPlayerReady(draftId, Player.GUEST);
                    playerType = Player.GUEST
                }
                socket.nsp
                    .in(roomHost)
                    .in(roomGuest)
                    .in(roomSpec)
                    .emit("player_joined", {name: message.name, playerType});
                fn({
                    ...draftsStore.getDraftViewsOrThrow(draftId).getDraftForPlayer(playerType),
                    yourPlayerType: playerType
                });
            });

            socket.on("act", Listeners.actListener(draftsStore, draftId, validateAndApply, socket, roomHost, roomGuest, roomSpec));

            socket.on('disconnect', function () {
                console.log('Got disconnect! draftId: ' + draftId);
            });
        });

        const httpServer = server.listen(port, () => {
            console.log("listening on *:" + port);
        });
        const httpServerAddr = httpServer.listen().address();


        function validateAndApply(draftId: string, message: DraftEvent): ValidationId[] {
            return validator.validateAndApply(draftId, message);
        }

        return {httpServer, httpServerAddr, io};
    }
};