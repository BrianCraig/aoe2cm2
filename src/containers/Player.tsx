import {IStoreState} from '../types';
import {connect} from 'react-redux';
import Player from "../components/Player";


export function mapStateToProps(state: IStoreState) {
    return {
        events: state.events,
        nextAction: state.nextAction
    };
}

export function mapDispatchToProps() {
    return {};
}

export default connect(mapStateToProps, mapDispatchToProps)(Player);