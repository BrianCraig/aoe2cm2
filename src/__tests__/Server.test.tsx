import {default as io} from "socket.io-client"
import {DraftServer} from "../DraftServer";
import request from "request"
import Player from "../models/Player";
import {IDraftConfig} from "../models/IDraftConfig";
import getPort from "get-port";

let hostSocket: any;
let clientSocket: any;
let httpServer: any;
let httpServerAddr: any;
let ioServer: any;
let draftId: string;

beforeAll((done) => {
    getPort().then((port: number) => {
        console.log("Got port: " + port);
        const serve = DraftServer.serve(port);
        httpServer = serve.httpServer;
        httpServerAddr = serve.httpServerAddr;
        ioServer = serve.io;
        done();
    });
});

afterAll((done) => {
    ioServer.close();
    httpServer.close();
    done();
});

beforeEach((done) => {
    request(`http://[${httpServerAddr.address}]:${httpServerAddr.port}/preset/new`, (error, response, body) => {
        const draftIdContainer: { draftId: string } = JSON.parse(body);
        draftId = draftIdContainer.draftId;
        hostSocket = io.connect(`http://[${httpServerAddr.address}]:${httpServerAddr.port}`, {
            query: {draftId: draftId},
            reconnectionDelay: 0,
            forceNew: true,
            transports: ['websocket'],
        });
        hostSocket.on('connect', () => {
            done();
        });
        clientSocket = io.connect(`http://[${httpServerAddr.address}]:${httpServerAddr.port}`, {
            query: {draftId: draftId},
            reconnectionDelay: 0,
            forceNew: true,
            transports: ['websocket'],
        });
        clientSocket.on('connect', () => {
            done();
        });

    });

});

afterEach((done) => {
    if (hostSocket.connected) {
        hostSocket.disconnect();
    }
    if (clientSocket.connected) {
        clientSocket.disconnect();
    }
    done();
});

it('successful join gets a draft config', (done) => {
    hostSocket.emit('join', {name: 'Saladin'}, (data: IDraftConfig) => {
        expect(data).toMatchSnapshot();
        done();
    });
});


it('should send player_joined when player joins', (done) => {
    hostSocket.once("player_joined", (message: any) => {
        expect(message.name).toBe('Saladin');
        expect(message.playerType).toBe(Player.HOST);

        hostSocket.once("player_joined", (message: any) => {
            expect(message.name).toBe('Barbarossa');
            expect(message.playerType).toBe(Player.GUEST);
        });
        clientSocket.once("player_joined", (message: any) => {
            expect(message.name).toBe('Barbarossa');
            expect(message.playerType).toBe(Player.GUEST);
            done();
        });
        clientSocket.emit('join', {name: 'Barbarossa'}, () => {
        });
    });
    hostSocket.emit('join', {name: 'Saladin'}, () => {
    });
});