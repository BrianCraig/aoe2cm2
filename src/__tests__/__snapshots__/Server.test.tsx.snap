// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`successful join gets a draft config 1`] = `
Object {
  "events": Array [],
  "guestReady": false,
  "hostReady": true,
  "language": "en-GB",
  "nameGuest": "â€¦",
  "nameHost": "Saladin",
  "nextAction": 0,
  "preset": Object {
    "encodedCivilisations": "7fffffff",
    "name": "Simple Preset",
    "turns": Array [
      Object {
        "action": "BAN",
        "exclusivity": "NONEXCLUSIVE",
        "hidden": false,
        "parallel": false,
        "player": "HOST",
      },
      Object {
        "action": "BAN",
        "exclusivity": "NONEXCLUSIVE",
        "hidden": false,
        "parallel": false,
        "player": "GUEST",
      },
      Object {
        "action": "PICK",
        "exclusivity": "NONEXCLUSIVE",
        "hidden": false,
        "parallel": false,
        "player": "GUEST",
      },
      Object {
        "action": "PICK",
        "exclusivity": "NONEXCLUSIVE",
        "hidden": false,
        "parallel": false,
        "player": "HOST",
      },
    ],
  },
  "yourPlayerType": "HOST",
}
`;
